File: code_cpp/self pased DSA/Tree/14-Construct-Quad-Tree.cpp
/*
Given a n * n matrix grid of 0's and 1's only. We want to represent the grid with a Quad-Tree.

Return the root of the Quad-Tree representing the grid.

Notice that you can assign the value of a node to True or False when isLeaf is False, and both are accepted in the answer.

A Quad-Tree is a tree data structure in which each internal node has exactly four children. Besides, each node has two attributes:

val: True if the node represents a grid of 1's or False if the node represents a grid of 0's.
isLeaf: True if the node is leaf node on the tree or False if the node has the four children.
class Node {
    public boolean val;
    public boolean isLeaf;
    public Node topLeft;
    public Node topRight;
    public Node bottomLeft;
    public Node bottomRight;
}
We can construct a Quad-Tree from a two-dimensional area using the following steps:

If the current grid has the same value (i.e all 1's or all 0's) set isLeaf True and set val to the value of the grid and set the four children to Null and stop.
If the current grid has different values, set isLeaf to False and set val to any value and divide the current grid into four sub-grids as shown in the photo.
Recurse for each of the children with the proper sub-grid.
*/

/*
// Definition for a QuadTree node.
class Node {
public:
    bool val;
    bool isLeaf;
    Node* topLeft;
    Node* topRight;
    Node* bottomLeft;
    Node* bottomRight;
    
    Node() {
        val = false;
        isLeaf = false;
        topLeft = NULL;
        topRight = NULL;
        bottomLeft = NULL;
        bottomRight = NULL;
    }
    
    Node(bool _val, bool _isLeaf) {
        val = _val;
        isLeaf = _isLeaf;
        topLeft = NULL;
        topRight = NULL;
        bottomLeft = NULL;
        bottomRight = NULL;
    }
    
    Node(bool _val, bool _isLeaf, Node* _topLeft, Node* _topRight, Node* _bottomLeft, Node* _bottomRight) {
        val = _val;
        isLeaf = _isLeaf;
        topLeft = _topLeft;
        topRight = _topRight;
        bottomLeft = _bottomLeft;
        bottomRight = _bottomRight;
    }
};
*/

class Solution {
public:
    Node* constructQuadTree(vector<vector<int>>& grid, int rowStart, int rowEnd, int colStart, int colEnd){
        if(rowStart>rowEnd || colStart>colEnd){
            return NULL;
        }

        bool isLeaf=true;
        int val = grid[rowStart][colStart];
        // check if all the values of grid are same or not
        for(int i=rowStart; i<=rowEnd; i++){
            for(int j=colStart; j<=colEnd; j++){
                if(grid[i][j]!=val){
                    isLeaf=false;
                    break;
                }
            }
            if(isLeaf==false) break;
        }
        // if all values are same then this a leaf node with value=val
        if(isLeaf==true){
            return new Node(val, true);
        }

        // all values are not same then split the grid into four parts
        int rowMid = (rowStart+rowEnd)/2;
        int colMid = (colStart+colEnd)/2;

        Node* topLeft = constructQuadTree(grid,rowStart ,rowMid, colStart, colMid);
        Node* topRight = constructQuadTree(grid,rowStart ,rowMid, colMid+1, colEnd);
        Node* bottomLeft = constructQuadTree(grid,rowMid+1 ,rowEnd, colStart, colMid);
        Node* bottomRight = constructQuadTree(grid,rowMid+1 ,rowEnd, colMid+1, colEnd);
        return new Node(false, false, topLeft, topRight, bottomLeft, bottomRight);
    }


    Node* construct(vector<vector<int>>& grid) {
        int n = grid.size();
        return constructQuadTree(grid, 0, n-1, 0, n-1);
    }
};

File: code_java/self pased DSA/Tree/14-Construct-Quad-Tree.java
/**
 * Problem: Construct Quad Tree from a 2D Grid
 * ---------------------------------------------
 * Given a 2D grid of 0s and 1s, construct a Quad Tree representing the grid.
 * Each node in the Quad Tree represents a region of the grid.
 *
 * Definition of Quad Tree Node:
 * - A leaf node represents a region with all values the same (all 0s or all 1s).
 * - A non-leaf node subdivides the region into four equal parts: topLeft, topRight, bottomLeft, bottomRight.
 *
 * Task:
 * Build a Quad Tree from the grid where each node corresponds to a region.
 * Leaf nodes hold the value (true for 1, false for 0) and mark isLeaf = true.
 * Non-leaf nodes have isLeaf = false and references to four children.
 *
 * Approach: Divide and Conquer (Recursive)
 * ---------------------------------------------
 * - Check if all values in the current sub-grid are the same.
 *   - If yes, create a leaf node with that value.
 *   - Else, split the grid into four equal quadrants and recursively build nodes for each quadrant.
 * - Combine the four child nodes into a parent node.
 *
 * Why it works:
 * - Quad Tree structure naturally represents spatial partitioning.
 * - Recursion breaks down the problem into smaller subproblems until uniform regions are found.
 * - This guarantees minimal tree size since uniform regions are combined into single leaf nodes.
 *
 * Steps to solve:
 * 1. Define base case: If sub-grid is uniform, create leaf node.
 * 2. Else, divide grid into four parts:
 *    topLeft, topRight, bottomLeft, bottomRight.
 * 3. Recursively construct Quad Trees for these four parts.
 * 4. Create a parent node with these four children and isLeaf = false.
 *
 * Time Complexity:
 * Worst case O(N^2 log N) where N is grid length.
 * - Checking uniformity in a sub-grid is O(k^2) for sub-grid size k.
 * - Recursion depth ~ log N (halving grid each step).
 *
 * Space Complexity:
 * O(N^2) for the recursion stack and output tree in worst case.
 */

class Node {
    public boolean val;
    public boolean isLeaf;
    public Node topLeft;
    public Node topRight;
    public Node bottomLeft;
    public Node bottomRight;

    public Node() {
        this.val = false;
        this.isLeaf = false;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }

    public Node(boolean val, boolean isLeaf) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }

    public Node(boolean val, boolean isLeaf, Node topLeft, Node topRight, Node bottomLeft, Node bottomRight) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = topLeft;
        this.topRight = topRight;
        this.bottomLeft = bottomLeft;
        this.bottomRight = bottomRight;
    }
}

public class Solution {

    private Node constructQuadTree(int[][] grid, int rowStart, int rowEnd, int colStart, int colEnd) {
        if (rowStart > rowEnd || colStart > colEnd) {
            return null;
        }

        // Check if all values in current sub-grid are same
        boolean isLeaf = true;
        int val = grid[rowStart][colStart];
        outerLoop:
        for (int i = rowStart; i <= rowEnd; i++) {
            for (int j = colStart; j <= colEnd; j++) {
                if (grid[i][j] != val) {
                    isLeaf = false;
                    break outerLoop;
                }
            }
        }

        if (isLeaf) {
            // Leaf node represents uniform grid area
            return new Node(val == 1, true);
        }

        // Otherwise, split grid into four equal parts and recurse
        int rowMid = (rowStart + rowEnd) / 2;
        int colMid = (colStart + colEnd) / 2;

        Node topLeft = constructQuadTree(grid, rowStart, rowMid, colStart, colMid);
        Node topRight = constructQuadTree(grid, rowStart, rowMid, colMid + 1, colEnd);
        Node bottomLeft = constructQuadTree(grid, rowMid + 1, rowEnd, colStart, colMid);
        Node bottomRight = constructQuadTree(grid, rowMid + 1, rowEnd, colMid + 1, colEnd);

        // Parent node with four children, not a leaf
        return new Node(false, false, topLeft, topRight, bottomLeft, bottomRight);
    }

    public Node construct(int[][] grid) {
        int n = grid.length;
        return constructQuadTree(grid, 0, n - 1, 0, n - 1);
    }
}